presto-main/src/main/java/com/facebook/presto/server/protocol/Query.java

```diff
@@ -383,26 +390,56 @@ private synchronized Optional<QueryResults> getCachedResult(long token)
+
+            if (originalBeforeRetryQueryId == null && previousQueryTopLevelPlanHash == null) {
+                originalBeforeRetryQueryId = queryId;
+                previousQueryTopLevelPlanHash = currentQueryTopLevelPlanHash;
+            }
         }
+        else if (queryManager.getQueryRetryCount(queryId) == 1 && retryQueryWithHistoryBasedOptimizationEnabled(session)) {
+            CanonicalPlanWithInfo currentQueryTopLevelPlanHash = queryManager.getFullQueryInfo(queryId).getPlanCanonicalInfo().get(0);
+            if (previousQueryTopLevelPlanHash.equals(currentQueryTopLevelPlanHash)) {
+                queryManager.failQuery(queryId, new PrestoException(GENERIC_INTERNAL_ERROR, "Since the plan hashes did not change, your retry query will not execute"));
+                return queryResults;
+            }
 
-        // check if we have exceeded the global limit
-        retryCircuitBreaker.incrementFailure();
-        if (!retryCircuitBreaker.isRetryAllowed() || hasProducedResult) {
             return queryResults;
         }
+        else {
+            if (queryResults.getError() == null || !queryResults.getError().isRetriable()) {
+                return queryResults;
+            }
 
-        // check if we have exceeded the local limit
-        if (queryManager.getQueryRetryCount(queryId) >= getQueryRetryLimit(session) ||
-                queryManager.getQueryInfo(queryId).getQueryStats().getExecutionTime().toMillis() > getQueryRetryMaxExecutionTime(session).toMillis()) {
-            return queryResults;
+            // check if we have exceeded the global limit
+            retryCircuitBreaker.incrementFailure();
+            if (!retryCircuitBreaker.isRetryAllowed() || hasProducedResult) {
+                return queryResults;
+            }
+
+            // check if we have exceeded the local limit
+            if (queryManager.getQueryRetryCount(queryId) >= getQueryRetryLimit(session) ||
+                    queryManager.getQueryInfo(queryId).getQueryStats().getExecutionTime().toMillis() > getQueryRetryMaxExecutionTime(session).toMillis()) {
+                return queryResults;
+            }
+
+            // no support for transactions
+            if (session.getTransactionId().isPresent() &&
+                    !transactionManager.getOptionalTransactionInfo(session.getRequiredTransactionId()).map(TransactionInfo::isAutoCommitContext).orElse(true)) {
+                return queryResults;
+            }
         }
 
-        // no support for transactions
-        if (session.getTransactionId().isPresent() &&
-                !transactionManager.getOptionalTransactionInfo(session.getRequiredTransactionId()).map(TransactionInfo::isAutoCommitContext).orElse(true)) {
-            return queryResults;
+        // no more retries left
+        if (!underOneRetryCount) {
+            originalBeforeRetryQueryId = null;
```

Comment:
Do these two originalBeforeRetryQueryId and previousQueryTopLevel ... need to be member variables ?

Reply:
@agrawaldevesh just previousQueryTopLevel is stored in the event of a retry, this way we can quickly compare the plan hashes